# 모의고사 1-2 통계분석 Q1~5

In [1]:
#---------------------------------------------------------------------------------------
# Q1) FIFA데이터에서 각 선수의 키는 Heghit변수에 피트와 인치로 입력되어 있습니다. 
#      이를 cm로 변환하여 새로운 변수 Height_cm을 생성하시오. 
#      ( “ ' ” 앞의 숫자는 피트이며, “ ' ” 뒤의 숫자는 인치, 1피트 = 30cm, 1인치 = 2.5cm)
#---------------------------------------------------------------------------------------

In [2]:
import pandas as pd
import numpy as np
df_fifa = pd.read_csv("FIFA.csv", encoding="cp949")
df_fifa

ID                Name  Age Nationality  Overall  \
0      158023            L. Messi   31   Argentina       94   
1       20801   Cristiano Ronaldo   33    Portugal       94   
2      190871           Neymar Jr   26      Brazil       92   
3      193080              De Gea   27       Spain       91   
4      192985        K. De Bruyne   27     Belgium       91   
...       ...                 ...  ...         ...      ...   
16637  238813        J. Lundstram   19     England       47   
16638  243165  N. Christoffersson   19      Sweden       47   
16639  241638           B. Worman   16     England       47   
16640  246268      D. Walker-Rice   17     England       47   
16641  246269           G. Nugent   16     England       46   

                      Club Preferred_Foot       Work_Rate Position  \
0             FC Barcelona           Left  Medium/ Medium       RF   
1                 Juventus          Right       High/ Low       ST   
2      Paris Saint-Germain          Right    High/ Medium       LW   
3        Manchester United          Right  Medium/ Medium       GK   
4          Manchester City          Right      High/ High      RCM   
...                    ...            ...             ...      ...   
16637      Crewe Alexandra          Right  Medium/ Medium       CM   
16638       Trelleborgs FF          Right  Medium/ Medium       ST   
16639     Cambridge United          Right  Medium/ Medium       ST   
16640      Tranmere Rovers          Right  Medium/ Medium       RW   
16641      Tranmere Rovers          Right  Medium/ Medium       CM   

       Jersey_Number  Contract_Valid_Until Height  Weight_lb  Release_Clause  \
0                 10                  2021    5'7        159          226500   
1                  7                  2022    6'2        183          127100   
2                 10                  2022    5'9        150          228100   
3                  1                  2020    6'4        168          138600   
4                  7                  2023   5'11        154          196400   
...              ...                   ...    ...        ...             ...   
16637             22                  2019    5'9        134             143   
16638             21                  2020    6'3        170             113   
16639             33                  2021    5'8        148             165   
16640             34                  2019   5'10        154             143   
16641             33                  2019   5'10        176             165   

        Value  Wage  
0      110500   565  
1       77000   405  
2      118500   290  
3       72000   260  
4      102000   355  
...       ...   ...  
16637      60     1  
16638      60     1  
16639      60     1  
16640      60     1  
16641      60     1  

[16642 rows x 16 columns]

In [3]:
df_fifa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16642 entries, 0 to 16641
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   ID                    16642 non-null  int64 
 1   Name                  16642 non-null  object
 2   Age                   16642 non-null  int64 
 3   Nationality           16642 non-null  object
 4   Overall               16642 non-null  int64 
 5   Club                  16642 non-null  object
 6   Preferred_Foot        16642 non-null  object
 7   Work_Rate             16642 non-null  object
 8   Position              16642 non-null  object
 9   Jersey_Number         16642 non-null  int64 
 10  Contract_Valid_Until  16642 non-null  int64 
 11  Height                16642 non-null  object
 12  Weight_lb             16642 non-null  int64 
 13  Release_Clause        16642 non-null  int64 
 14  Value                 16642 non-null  int64 
 15  Wage                  16642 non-null

In [4]:
df_fifa["Height_cm"] = df_fifa["Height"]

In [5]:
df_fifa["Height_cm"] = df_fifa["Height_cm"].map(lambda x: int(x.split("'")[0])*30 + int(x.split("'")[1])*2.5)
df_fifa["Height_cm"]

0        167.5
1        185.0
2        172.5
3        190.0
4        177.5
         ...  
16637    172.5
16638    187.5
16639    170.0
16640    175.0
16641    175.0
Name: Height_cm, Length: 16642, dtype: float64

# Q1 끝

In [6]:
#---------------------------------------------------------------------------------------
# Q2) 포지션을 의미하는 Position변수를 아래 표를 참고하여 “Forward”, “Midfielder”, 
#      “Defender”, “GoalKeeper”로 재범주화하고, factor형으로 변환하여 Position_Class 
#       라는 변수를 생성하고 저장하시오.
#---------------------------------------------------------------------------------------

In [7]:
df_fifa["Position"].value_counts()

ST     1924
GK     1900
CB     1646
CM     1276
LB     1223
RB     1199
RM     1022
LM     1012
CAM     876
CDM     873
RCB     615
LCB     595
RCM     355
LCM     353
LW      338
RW      325
RDM     228
LDM     224
LS      184
RS      179
RWB      84
LWB      76
CF       66
RAM      21
LAM      20
LF       15
RF       13
Name: Position, dtype: int64

In [8]:
Forward = ["LS","ST","RS","LW","LF","CF","RF","RW"]
Midfielder = ["LAM","CAM","RAM","LM","LCM","CM","RCM","RM"]
Defender = ["LWB","LDM","CDM","RDM","RWB","LB","LCB","CB","RCB","RB"]
Goalkeeper = ["GK"]

In [9]:
df_fifa.loc[df_fifa["Position"].isin(Forward),"Position_Class"] = "Forward"
df_fifa.loc[df_fifa["Position"].isin(Midfielder),"Position_Class"] = "Midfielder"
df_fifa.loc[df_fifa["Position"].isin(Defender),"Position_Class"] = "Defender"
df_fifa.loc[df_fifa["Position"].isin(Goalkeeper),"Position_Class"] = "Goalkeeper"

In [18]:
df_fifa["Position_Class"].value_counts()

Defender      6763
Midfielder    4935
Forward       3044
Goalkeeper    1900
Name: Position_Class, dtype: int64

## Q2 끝

In [14]:
#---------------------------------------------------------------------------------------
# Q3) 새로 생성한 Position_Class 변수의 각 범주에 따른 Value(선수의 시장가치)의 
#      평균값의 차이를 비교하는 일원배치 분산분석을 수행하고 결과를 해석하시오. 
#      그리고 평균값의 차이가 통계적으로 유의하다면 사후검정을 수행하고 설명하시오.
#---------------------------------------------------------------------------------------

In [17]:
from statsmodels.stats.anova import anova_lm
from statsmodels.formula.api import ols

anova_lm(ols("Value ~ C(Position_Class)", data = df_fifa).fit())

df        sum_sq       mean_sq          F  \
C(Position_Class)      3.0  4.081181e+09  1.360394e+09  41.873906   
Residual           16638.0  5.405330e+11  3.248786e+07        NaN   

                         PR(>F)  
C(Position_Class)  5.988667e-27  
Residual                    NaN

## Q3-1
Value 값을 Position_Class를 요인으로 ANOVA 분석 결과,

F 통계값에 대한 P-value가 유의수준 0.05에 비해 현저히 낮으므로

귀무가설인 mean(Position_Class == Forward) = mean(Position_Class == Midfielder) = ... = mean(Position_Class == GoalKeeper)를 기각하여,

Position_Class 요인에 따라 적어도 한 쌍은 평균이 통계적으로 다르다라고 말할 수 있다.

In [20]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd
pairwise_tukeyhsd(df_fifa["Value"], df_fifa["Position_Class"]).summary()

<class 'statsmodels.iolib.table.SimpleTable'>

## Q3-2
Tukey HSD로 요인 Position_Class 에 대한 쌍검정을 했을때,

Forward - Midfeilder 쌍을 제외하고, 모두 meandiff 값이 HSD 범위를 벗어났으므로, Value 평균이 0.05의 유의수준에 대해 다르다고 말할 수 있다.

Forward - Midfeilder 쌍은 meandiff 값이 HSD 범위 내에 있으므로, Value 평균이 다르다고 말할 수 없다.

In [21]:
#---------------------------------------------------------------------------------------
# Q4) Preferred Foot(주로 사용하는 발)과 Position_Class(재범주화 된 포지션)변수에 
#      따라 Value(이적료)의 차이가 있는지를 알아보기 위해 이원배치분산분석을 
#      수행하고 결과를 해석하시오.
#--------------------------------------------------------------------------------------

In [22]:
anova_lm(ols("Value ~ Preferred_Foot * Position_Class", data = df_fifa).fit())

df        sum_sq       mean_sq          F  \
Preferred_Foot                     1.0  1.460850e+08  1.460850e+08   4.500734   
Position_Class                     3.0  4.087367e+09  1.362456e+09  41.975902   
Preferred_Foot:Position_Class      3.0  4.736156e+08  1.578719e+08   4.863874   
Residual                       16634.0  5.399071e+11  3.245805e+07        NaN   

                                     PR(>F)  
Preferred_Foot                 3.389505e-02  
Position_Class                 5.151392e-27  
Preferred_Foot:Position_Class  2.207249e-03  
Residual                                NaN

## Q4-1
주 요인인, Preferred_Foot, Position_Class 에 대해 F 값에 대한 P-value가 유의수준 0.05에 대해 현저히 낮으므로,

귀무가설을 기각하여, 주 요인에 따라 Value 평균이 다르다고 말할 수 있다.

또한 교호작용 요인인, Preferred_Foot * Position_Class 에 대해서도 F값에 대한 P-value가 유의수준 0.05에 비해 낮기 때문에,

귀무가설을 기각하여, 두 요인의 교호작용에 대해서도 Value 평균이 달라진다고 말할 수 있다.

In [25]:
df_fifa["Preferred_Foot_Position_Class"] = df_fifa["Preferred_Foot"] + "_" + df_fifa["Position_Class"]
df_fifa["Preferred_Foot_Position_Class"].value_counts()

Right_Defender      4838
Right_Midfielder    3747
Right_Forward       2530
Left_Defender       1925
Right_Goalkeeper    1707
Left_Midfielder     1188
Left_Forward         514
Left_Goalkeeper      193
Name: Preferred_Foot_Position_Class, dtype: int64

In [26]:
pairwise_tukeyhsd(df_fifa["Value"], df_fifa["Preferred_Foot_Position_Class"]).summary()

<class 'statsmodels.iolib.table.SimpleTable'>

## Q4-2
두 요인의 조합인 Preferred_Foot * Position_Class 에 대해 Tukey HSD 구간 검정을 했을때,

reject가 True인 쌍은 서로 value 평균이 다르다고 말할 수 있고 (유의수준 0.05)

reject가 False인 쌍은 서로 value 평균이 다르다고 말할 수 없다 (유의수준 0.05)

In [27]:
#---------------------------------------------------------------------------------------
# Q5)Age, Overall, Wage, Height_cm, Weight_lb 가 Value에 영향을 미치는지 
#     알아보는 회귀분석을 단계적 선택법을 사용하여 수행하고 결과를 해석하시오. 
#---------------------------------------------------------------------------------------

In [47]:
# 1, 상수항
model = ols("Value ~ 1", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Age", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Overall", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Height_cm", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Weight_lb", data = df_fifa)
print(model.fit().bic)

335205.1558283197
335117.3444646266
326827.08244640735
312652.15076842374
335214.8160892864
335182.4193721193


In [50]:
# 2, Wage 추가
model = ols("Value ~ Wage", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ 1", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage + Age", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage + Overall", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage + Height_cm", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage + Weight_lb", data = df_fifa)
print(model.fit().bic)

312652.15076842374
335205.1558283197
312485.4822709994
310815.57190793526
312646.37395881966
312652.42481762206


In [53]:
# 3, Wage + Overall
model = ols("Value ~ Wage + Overall", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Overall", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage + Overall + Age", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage + Overall + Height_cm", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage + Overall + Weight_lb", data = df_fifa)
print(model.fit().bic)

310815.57190793526
312652.15076842374
326827.08244640735
309204.04792635
310793.6671959044
310731.5756807156


In [55]:
# 4, Wage + Overal + Age
model = ols("Value ~ Wage + Overall + Age", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Overall + Age", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage + Age", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage + Overall", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage + Overall + Age + Height_cm", data = df_fifa)
print(model.fit().bic)
model = ols("Value ~ Wage + Overall + Age + Weight_lb", data = df_fifa)
print(model.fit().bic)

309204.04792635
325109.2977464062
312485.4822709994
310815.57190793526
309206.2599529369
309206.4746432461


In [62]:
# 5, Wage + Overal + Age 의 BIC 값보다 작아지는 벌점이 없으므로, 변수 추가 또는 삭제 종료
# 최종 모델은 Value ~ Wage + Overal + Age

model = ols("Value ~ Wage + Overall + Age", data = df_fifa)
model.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Value   R-squared:                       0.791
Model:                            OLS   Adj. R-squared:                  0.791
Method:                 Least Squares   F-statistic:                 2.096e+04
Date:                Sat, 18 Dec 2021   Prob (F-statistic):               0.00
Time:                        13:00:19   Log-Likelihood:            -1.5458e+05
No. Observations:               16642   AIC:                         3.092e+05
Df Residuals:                   16638   BIC:                         3.092e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -1.018e+04    228.843    -44.466      0.000   -1.06e+04   -9727.191
Wage         184.1522      1.128    163.265      0.000     181.941     186.363
Overall      241.4204      4.002     60.318      0.000     233.575     249.266
Age         -203.1997      4.925    -41.261      0.000    -212.853    -193.547
==============================================================================
Omnibus:                    17093.962   Durbin-Watson:                   1.409
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          3530710.483
Skew:                           4.667   Prob(JB):                         0.00
Kurtosis:                      73.743   Cond. No.                         813.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Q5
최종 모델은 Value =  -1.018e+04 + 184.1522 * Wage + 241.4204 * Overall + -203.1997 * Age 이며,

수정된 R2 결정계수는 0.791로 모델이 약 79%를 설명한다고 할 수 있다.

각각의 coef는 P value가 유의수준보다 낮으므로 (0.05) 통계적으로 유의하다.

Wage 가 증가할 때 value 가 증가하며, Overall 이 증가할 때도 value 는 증가한다.

Age 가 증가할 때는 value 는 감소한다.

또한 coef 절대값이 Overall 이 가장 크므로, value에 대한 영향력이 가장 크다고 할 수 있다.